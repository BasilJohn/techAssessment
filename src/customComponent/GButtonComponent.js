/**
 * @format
 * @flow strict-local
 */
import React from 'react';
import { NavigationScreenProps } from '@react-navigation/native';
import {
  View, Text, TouchableOpacity, Dimensions, StyleSheet
} from 'react-native';
import * as colorConstants from '../constants/colorConstants';

type buttonProps = NavigationScreenProps &{
    containerStyle?:Object,
    textStyle?:Object
  };

const buttonComponent = (props: buttonProps) => (
  <View style={[styles.buttonStyle, props.containerStyle]}>
    <TouchableOpacity style={styles.button}>
      <Text testID="btnText" style={[styles.buttonTextStyle, props.textStyle]}>Press me</Text>
    </TouchableOpacity>
  </View>
);

const styles = StyleSheet.create({
  buttonStyle: {
    width: '90%',
    alignSelf: 'center',
    height: Dimensions.get('window').height * 0.07,
    borderRadius: 10,
    marginBottom: 10,
  },
  buttonTextStyle: {
    color: colorConstants.blueText,
    textAlign: 'center',
  },
  button: {
    height: '100%',
    width: '100%',
    justifyContent: 'center',
  },
});

export default buttonComponent;
