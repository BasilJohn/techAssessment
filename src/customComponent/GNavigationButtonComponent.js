/**
 * @format
 * @flow strict-local
 */
import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  Dimensions,
  StyleSheet,
} from 'react-native';
import * as colorConstants from '../constants/colorConstants';
import {NavigationScreenProps} from '@react-navigation/native';
type validatorProps = NavigationScreenProps & {
  objectOne?: Object,
  objectTwo?: Object,
  operation?: Function,
};

const navigationButtonComponent = (props: validatorProps) => {
  return (
    <View style={styles.buttonContainer}>
      <TouchableOpacity
        onPress={() => props.navProps.navigation.navigate(props.screenName)}
        style={styles.button}>
        <Text style={styles.textStyle}>{props.title}</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  buttonContainer: {
    width: '90%',
    height: Dimensions.get('window').height * 0.1,
    backgroundColor: colorConstants.navigationButtonColor,
    alignSelf: 'center',
    marginBottom: 10,
  },
  button: {
    height: '100%',
    width: '100%',
    justifyContent: 'center',
  },
  textStyle: {
    textAlign: 'center',
  },
});

export default navigationButtonComponent;
