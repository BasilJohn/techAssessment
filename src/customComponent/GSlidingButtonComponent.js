/**
 * @format
 * @flow strict-local
 */
import * as React from 'react';
import {
  Animated,
  View,
  StyleSheet,
  PanResponder,
  Dimensions,
  Text,
} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';

import * as colorConstants from '../constants/colorConstants';

type Props = {
  endAction: () => void,
};
type State = {
  bgColor: number
};
class SlidingButton extends React.Component<Props, State> {
  deviceWidth = Dimensions.get('window').width;

  calculatedWidth = 0;

  animatedWidth = new Animated.Value(0);

  position = new Animated.ValueXY();

  bgPanHandler = PanResponder.create({
    onMoveShouldSetPanResponder: () => true,
    onPanResponderGrant: () => true,
    onPanResponderMove: (event, gesture) => {
      const runningPC = (gesture.dx / this.deviceWidth) * 100;
      if (runningPC < 0) {
        Animated.timing(this.animatedWidth, {
          toValue: 20,
          duration: 0,
          useNativeDriver: false,
        }).start();
      } else if (runningPC > 71.5) {
        Animated.timing(this.animatedWidth, {
          toValue: 100,
          duration: 0,
          useNativeDriver: false,
        }).start();
        this.props.endAction();
      } else {
        Animated.timing(this.animatedWidth, {
          toValue: 20 + runningPC,
          duration: 0,
          useNativeDriver: false,
        }).start();
      }
      const bgColorRunning = this.animatedWidth._value === 0 ? 0 : runningPC;
      this.setState({ bgColor: bgColorRunning });
    },
    onPanResponderRelease: (event, gesture) => {
      const runningPC = (gesture.dx / this.deviceWidth) * 100;
      if (runningPC < 71.5) {
        Animated.timing(this.animatedWidth, {
          toValue: 20,
          duration: 0,
          useNativeDriver: false,
        }).start();
      }
      const bgColorRunning = this.animatedWidth._value === 0 ? 0 : runningPC;
      this.setState({ bgColor: bgColorRunning });
      this.animatedWidth.flattenOffset();
    },
  });

  constructor(props:Object) {
    super(props);
    this.state = {
      bgColor: 0
    };
  }

  render(): React.Node {
    return (
      <View style={styles.container}>
        <View style={[styles.iconContainer]}>
          <View style={[styles.textContainer]}>
            <Text testID="txtSlideText" style={[styles.slideTxt]}>
              Slide me to continue
            </Text>
          </View>
          <Animated.View
            {...this.bgPanHandler.panHandlers}
            style={[
              styles.button,
              {
                width:
                  this.state.bgColor === 0
                    ? '20%'
                    : `${this.animatedWidth._value}%`,
              },
            ]}
          >
            <View style={styles.icon}>
              <Icon name="diamond" size={20} style={styles.iconText} />
            </View>
          </Animated.View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  iconContainer: {
    width: '90%',
    borderRadius: 10,
    alignSelf: 'center',
    height: 60,
    borderWidth: 1,
    borderColor: colorConstants.dragButtonBorderColor,
    backgroundColor: colorConstants.dragButtonBGColor,
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 10,
    height: '100%',
    backgroundColor: colorConstants.dragColor,
  },
  icon: {
    height: '100%',
    width: Dimensions.get('window').width / 5.6,
    backgroundColor: colorConstants.darkButtonBG,
    alignSelf: 'flex-end',
    borderRadius: 10,
    justifyContent: 'center',
    overflow: 'hidden',
  },
  iconText: {
    textAlign: 'center',
    color: colorConstants.white,
  },
  slideTxt: {
    color: colorConstants.white,
    textAlign: 'center',
  },
  textContainer: {
    position: 'absolute',
    justifyContent: 'center',
    width: '100%',
    height: '100%',
  },
});

export default SlidingButton;
