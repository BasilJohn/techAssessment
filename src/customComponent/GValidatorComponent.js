/**
 * @format
 * @flow strict-local
 */
import React from 'react';
import {
  View,
  Text,
  StyleSheet,
} from 'react-native';
import { NavigationScreenProps } from '@react-navigation/native';

type validatorProps = NavigationScreenProps & {
  objectOne?: Object,
  objectTwo?: Object,
  operation?:Object
};

const validatorComponent = (props: validatorProps) => {
  return (
    <View style={styles.validatorContainer}>
      <Text style={styles.header}>{props.headerOne}</Text>
      <Text>{JSON.stringify(props.objectOne)}</Text>
      <Text style={styles.header}>{props.headerTwo}</Text>
      <Text>{JSON.stringify(props.objectTwo)}</Text>
      <Text style={styles.header}>{props.final}</Text>
      <Text>
        {JSON.stringify(props.operation(props.objectOne, props.objectTwo))}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  validatorContainer: {
    padding: 10,
  },
  header: {
    fontWeight: 'bold',
    margin: 5,
  },
});

export default validatorComponent;
