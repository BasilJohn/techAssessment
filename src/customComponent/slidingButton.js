/**
 * @format
 * @flow strict-local
 */
import * as React from 'react';
import {
  Animated,
  View,
  StyleSheet,
  PanResponder,
  Dimensions,
  Text,
  Easing
} from 'react-native';
import * as colorConstants from '../constants/colorConstants';
import Icon from 'react-native-vector-icons/FontAwesome';

type Props = {
  endAction: () => void,
};
type State = {
  bgColor: string,
  finalWidth: number,
};

class SlidingButton extends React.Component<Props, State> {
  state = {
    bgColor: colorConstants.white,
    finalWidth: 0,
  };
  deviceWidth = Dimensions.get('window').width;
  calculatedWidth = 0;
  animatedWidth = new Animated.Value(0);
  position = new Animated.ValueXY();
  bgPanHandler = PanResponder.create({
    onMoveShouldSetPanResponder: () => true,
    onPanResponderGrant: () => true,
    onPanResponderMove: (event, gesture) => {
      const runningPC = (gesture.dx / this.deviceWidth) * 100;
      if (runningPC < 0) {
        Animated.timing(this.animatedWidth, {
          toValue: 20,
          duration: 0,
          useNativeDriver: false,
        }).start();
      } else if (runningPC > 71.5) {
        Animated.timing(this.animatedWidth, {
          toValue: 100,
          duration: 0,
          useNativeDriver: false,
        }).start();
        this.props.endAction();
      } else {
        Animated.timing(this.animatedWidth, {
          toValue: 20 + runningPC,
          duration: 0,
          useNativeDriver: false,
        }).start();
      }
      let bgColorRunning =
        this.animatedWidth._value == 0 ? '#262A32' : 'rgba(124,243,249, 0.8)';
      this.setState({bgColor: bgColorRunning});
    },
    onPanResponderRelease: (event, gesture) => {
      const runningPC = (gesture.dx / this.deviceWidth) * 100;
      if (runningPC < 71.5) {
        Animated.timing(this.animatedWidth, {
          toValue: 20,
          duration: 0,
          useNativeDriver: false,
        }).start();
      }
      let bgColorRunning =
      this.animatedWidth._value == 0 ? '#262A32' : 'rgba(124,243,249, 0.8)';
       this.setState({bgColor: bgColorRunning});
      this.animatedWidth.flattenOffset();
    },
  });

  render(): React.Node {
    return (
      <View style={styles.container}>
        <View style={[styles.iconContainer]}>
          <View style={[styles.textContainer]}>
            <Text style={[styles.slideTxt]}>Slide me to continue</Text>
          </View>
          <Animated.View
            {...this.bgPanHandler.panHandlers}
            style={[
              styles.button,
              {
                width:
                  this.animatedWidth._value == 0
                    ? '20%'
                    : this.animatedWidth._value + '%',
              },
            ]}>
            <View style={styles.icon}>
              <Icon name="diamond" size={20} style={styles.iconText} />
            </View>
          </Animated.View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  iconContainer: {
    width: '90%',
    borderRadius: 10,
    alignSelf: 'center',
    height: 60,
    borderWidth: 1,
    borderColor: colorConstants.dragButtonBorderColor,
    backgroundColor: colorConstants.dragButtonBGColor,
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 10,
    height: '100%',
    backgroundColor: colorConstants.dragColor,
  },
  icon: {
    height: '100%',
    width: Dimensions.get('window').width / 5.6,
    backgroundColor: colorConstants.darkButtonBG,
    alignSelf: 'flex-end',
    borderRadius: 10,
    justifyContent: 'center',
    overflow: 'hidden',
  },
  iconText: {
    textAlign: 'center',
    color: colorConstants.white,
  },
  slideTxt: {
    color: colorConstants.white,
    textAlign: 'center',
  },
  textContainer: {
    position: 'absolute',
    justifyContent: 'center',
    width: '100%',
    height: '100%',
  },
});

export default SlidingButton;
