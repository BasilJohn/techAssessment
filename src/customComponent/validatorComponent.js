
/**
 * @format
 * @flow strict-local
 */
import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  Dimensions,
  StyleSheet,
} from 'react-native';
import * as colorConstants from '../constants/colorConstants';
import {NavigationScreenProps} from '@react-navigation/native';

type validatorProps = NavigationScreenProps & {
  objectOne?: Object,
  objectTwo?: Object,
  operation?:Function
};

const validatorComponent = (props: validatorProps) => {
  return (
    <View style={styles.validatorContainer}>
      <Text style={styles.header}>{props.headerOne}</Text>
      <Text>{JSON.stringify(props.objectOne)}</Text>
      <Text style={styles.header}>{props.headerTwo}</Text>
      <Text>{JSON.stringify(props.objectTwo)}</Text>
      <Text style={styles.header}>{props.final}</Text>
      <Text>
        {JSON.stringify(props.operation(props.objectOne, props.objectTwo))}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  validatorContainer: {
    padding: 10,
  },
  header: {
    fontWeight: 'bold',
    margin: 5,
  },
});

export default validatorComponent;
