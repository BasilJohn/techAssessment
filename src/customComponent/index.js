import GButtonComponent from './GButtonComponent';
import GSlidingButtonComponent from './GSlidingButtonComponent';
import GNavigationButtonComponent from './GNavigationButtonComponent';
import GValidatorComponent from './GValidatorComponent';

export {
  GButtonComponent, GSlidingButtonComponent, GNavigationButtonComponent, GValidatorComponent
};
