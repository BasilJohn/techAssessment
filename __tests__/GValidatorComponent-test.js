/**
 * @format
 */

import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { GValidatorComponent } from '../src/customComponent';

const operation = jest.fn();

test('Renders GButtonComponent Component Correctly', () => {
  const tree = renderer.create(<GValidatorComponent operation={operation} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Check prop value is being set', () => {
  const wrap = shallow(
    <GValidatorComponent operation={operation} />
  );
  wrap.setProps({ final: 'Combined Object' });
  expect(wrap.find('Text').getElements()[4].props.children).toEqual('Combined Object');
});
