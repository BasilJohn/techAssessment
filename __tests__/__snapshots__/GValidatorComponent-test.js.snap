// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders GButtonComponent Component Correctly 1`] = `
<View
  style={
    Object {
      "padding": 10,
    }
  }
>
  <Text
    style={
      Object {
        "fontWeight": "bold",
        "margin": 5,
      }
    }
  />
  <Text />
  <Text
    style={
      Object {
        "fontWeight": "bold",
        "margin": 5,
      }
    }
  />
  <Text />
  <Text
    style={
      Object {
        "fontWeight": "bold",
        "margin": 5,
      }
    }
  />
  <Text />
</View>
`;
